lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -1865,6 +1865,13 @@ RequirementCheck WitnessChecker::checkWitness(ValueDecl *requirement,
       std::make_pair(AccessScope::getPublic(), false));
 
   bool isSetter = false;
+  if (match.Witness->isAccessibleFrom(requiredAccessLevel.getDeclContext(),
+                                      true) &&
+      !match.Witness->isAccessibleFrom(requiredAccessLevel.getDeclContext(),
+                                       false)) {
+    return RequirementCheck(CheckKind::AccessStrict, requiredAccessLevel,
+                            isSetter);
+  }
   if (checkWitnessAccess(DC, requirement, match.Witness, &isSetter))
```

Comment:
You could sink your new check down into `checkWitnessAccess()` if you change the return type of `checkWitnessAccess()`. That would avoid repeating the `/*forConformance=*/true` check twice.